### PR TITLE
Profile restructuring

### PR DESCRIPTION
**Pull request by mkrabacher**
profile-contributions component created and implemented.

### CODE REVIEW

File: src/app/components/account/profile-contributions/profile-contributions.component.html [hunk 1]
```diff
@@ -1,31 +1,35 @@
-<h3>
-    Recent Authored Articles
-    <a *ngIf="displayedAuthoredArticles.length <= minDisplayNum && authoredArticles.length > minDisplayNum"
-       (click)="toggleAllAuthored()"
+<div *ngIf="displayedAuthoredArticles.length > 0">
+  <h3>
+      Recent Authored Articles
+      <a *ngIf="displayedAuthoredArticles.length <= minDisplayNum && authoredArticles.length > minDisplayNum"
+        (click)="toggleAllAuthored()"
+          target="_blank">(show all)</a>
```
[the remark this answers is not in the document]
Comment (mkrabacher) [reply src/app/components/account/profile-contributions/profile-contributions.component.html]: You're absolutely right. I assume that was either muscle memory and I just threw it in or I just copied it from some previous work where I wanted functionality that I'd already created. Either way I'm confident that I didn't put a whole lot of thought into the actual UI/UX and naming conventions of the template because I was assuming, maybe wrongly, that you would want to go over/dictate the look and use cases of the component and then we'd go from there. Judging from a light skimming of the following change requests it seems like they mostly stem from that assumption.

tl;dr The UI/UX that I wrote for this component was meant as a placeholder until more direction was received.

File: src/app/components/account/profile/profile.component.html [hunk 2]
```diff
@@ -1,24 +1,31 @@
-<section class="profile">
+<section>
```
[the remark this answers is not in the document]
Comment (mkrabacher) [reply src/app/components/account/profile/profile.component.html]: I saw the article you posted and I'll incorporate it into the template.

File: src/app/components/account/profile-contributions/profile-contributions.component.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+@Component({
+  selector: 'cos-profile-contributions',
+  templateUrl: './profile-contributions.component.html',
+  styleUrls: ['./profile-contributions.component.scss']
+})
+export class ProfileContributionsComponent implements OnInit, OnChanges {
+  @Input() loggedInUserId;
+  @Input() profileId;
+  @Input() minDisplayNum = 3;
```
[the remark this answers is not in the document]
Comment (mkrabacher) [reply src/app/components/account/profile-contributions/profile-contributions.component.ts]: Gosh, can't believe I didn't use that variable on those methods. That was silly. Did you mean to leave this set to 3 though? If you're are setting an @Input on the creation of the component doesn't that completely defeat the purpose?

File: src/app/components/account/profile-contributions/profile-contributions.component.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    this.editedArticlesMap = {};
+    this.authoredArticlesMap = {};
+    this.setContributionInfo();
+  }
+
+  watchAuthoredArticles() {
+    this.authoredArticles = [];
+    this.displayedAuthoredArticles = [];
+
+    this.articleSvc.getArticleRefsByAuthor(this.profileId).valueChanges().subscribe(previews => {
```
[the remark this answers is not in the document]
Comment (mkrabacher) [reply src/app/components/account/profile-contributions/profile-contributions.component.ts]: Man, this was helpful. I'm still struggling with the best way to work with subscriptions and I'm not sure I fully understand what the differences between valueChanges() and get() are, other than they return slightly different info, but your rewrite is way easier to understand. I hated all the subscriptions that were happening in those functions.

File: src/app/components/account/profile-contributions/profile-contributions.component.ts [hunk 5]
```diff
@@ -31,10 +33,9 @@ export class ProfileContributionsComponent implements OnInit, OnChanges {
   }
 
   watchAuthoredArticles() {
```
Comment (mkrabacher) [inline src/app/components/account/profile-contributions/profile-contributions.component.ts]: To elaborate on this a little more; something about your changes reopened a bug I remember having the first time around where this subscribe would end up happening three time on component initialization. And because of the asynchronous function call watchAuthoredArticles() gets called three times before the first getPreviewRefsByAuthor() returns anything, meaning that the article arrays are set to empty three times before anything is added. Then, the for(key in ArticlesMap) is called three times without ever resetting the the article arrays, thus ending up with ArticleArrays that look like [art1, art2, art1, art2, art1, art2]. Hope that makes sense. Happy to talk further about it.

File: src/app/components/account/profile-contributions/profile-contributions.component.ts [hunk 6]
```diff
@@ -33,62 +31,66 @@ export class ProfileContributionsComponent implements OnInit, OnChanges {
...
-              this.displayedAuthoredArticles.push(artPrev);
-            }
-          }
-        });
+    this.articleSvc.getPreviewRefsByAuthor(authorId).valueChanges().subscribe(previews => {
+      previews.forEach(preview => {
+        this.authoredArticlesMap[preview.articleId] = preview;
       });
+
+      for (const key in this.authoredArticlesMap) {
```
[the remark this answers is not in the document]
Comment (Methodician) [reply src/app/components/account/profile-contributions/profile-contributions.component.ts]: That seems weird because we're iterating over the keys in the map. I suppose it's plausible that one of the articles could be removed from a map during an iteration. Can't think of any other reason to do that.

File: src/app/components/account/profile/profile.component.scss [hunk 7]
```diff
@@ -1,31 +1,31 @@
...
 .profile {
-  max-width: 768px;
-  margin: auto;
   padding: 24px 24px 48px;
 
   &__title {
     color: $primary-color;
     text-align: center;
   }
+
```
Comment (Methodician) [inline src/app/components/account/profile/profile.component.scss]: Very nice.

File: src/app/components/account/profile-contributions/profile-contributions.component.ts (hunk 5 above)
Comment (Methodician) [reply src/app/components/account/profile-contributions/profile-contributions.component.ts]: Turns out this was caused by ngOnChanges. Since I couldn't understand why it was there I just removed it. Hoefully that doesn't mess anything up...

File: src/app/components/account/profile-contributions/profile-contributions.component.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Methodician) [reply src/app/components/account/profile-contributions/profile-contributions.component.ts]: Now that I read the rule you linked to it's making more sense. I've never had that happen but at least it's plausible.

File: src/app/components/account/profile-contributions/profile-contributions.component.ts [hunk 8]
```diff
@@ -43,16 +43,9 @@ export class ProfileContributionsComponent implements OnInit {
         this.authoredArticlesMap[preview.articleId] = preview;
       });
 
-      for (const key in this.authoredArticlesMap) {
-        if (this.authoredArticlesMap.hasOwnProperty(key)) {
+      this.authoredArticles = Object.values(this.authoredArticlesMap);
+      this.displayedAuthoredArticles = this.authoredArticles.slice(0, this.minDisplayNum);
```
Comment (mkrabacher) [inline src/app/components/account/profile-contributions/profile-contributions.component.ts]: so nice...
